src/bank.rs

```diff
@@ -335,91 +327,130 @@ impl Bank {
-            return Err(BankError::ModifiedContractId);
+            return Err(BankError::ModifiedContractId(pix as u8));
         }
         // For accounts unassigned to the contract, the individual balance of each accounts cannot decrease.
-        if tx.program_id != account.program_id && pre_tokens > account.tokens {
-            return Err(BankError::ExternalAccountTokenSpend);
+        if *tx_program_id != account.program_id && pre_tokens > account.tokens {
+            return Err(BankError::ExternalAccountTokenSpend(pix as u8));
         }
         if account.tokens < 0 {
-            return Err(BankError::ResultWithNegativeTokens);
+            return Err(BankError::ResultWithNegativeTokens(pix as u8));
         }
         Ok(())
     }
 
-    fn loaded_contract(&self, tx: &Transaction, accounts: &mut [Account]) -> bool {
+    fn loaded_contract(
+        &self,
+        tx_program_id: &Pubkey,
+        tx: &Transaction,
+        pix: usize,
+        accounts: &mut [&mut Account],
+    ) -> bool {
         let loaded_contracts = self.loaded_contracts.write().unwrap();
-        match loaded_contracts.get(&tx.program_id) {
+        match loaded_contracts.get(&tx_program_id) {
             Some(dc) => {
-                let mut infos: Vec<_> = (&tx.keys)
+                let mut infos: Vec<_> = (&tx.account_keys)
                     .into_iter()
                     .zip(accounts)
                     .map(|(key, account)| KeyedAccount { key, account })
                     .collect();
 
-                dc.call(&mut infos, &tx.userdata);
+                dc.call(&mut infos, tx.userdata(pix));
                 true
             }
             None => false,
         }
     }
-
+    fn with_subset<F, A>(accounts: &mut [Account], ixes: &[u8], func: F) -> A
+    where
+        F: Fn(&mut [&mut Account]) -> A,
+    {
+        let mut subset: Vec<&mut Account> = ixes
+            .iter()
+            .map(|ix| {
+                let ptr = &mut accounts[*ix as usize] as *mut Account;
+                // lifetime of this unsafe is only within the scope of the closure
+                // there is no way to reorder them without breaking borrow checker rules
+                unsafe { &mut *ptr }
+            }).collect();
+        func(&mut subset)
+    }
     /// Execute a transaction.
     /// This method calls the contract's process_transaction method and verifies that the result of
     /// the contract does not violate the bank's accounting rules.
     /// The accounts are committed back to the bank only if this function returns Ok(_).
-    fn execute_transaction(&self, tx: &Transaction, accounts: &mut [Account]) -> Result<()> {
-        let pre_total: i64 = accounts.iter().map(|a| a.tokens).sum();
-        let pre_data: Vec<_> = accounts
-            .iter_mut()
-            .map(|a| (a.program_id, a.tokens))
-            .collect();
+    fn execute_transaction(&self, tx: &Transaction, tx_accounts: &mut [Account]) -> Result<()> {
```

Comment:
This can be two functions, the original function renamed to `execute_instruction` and this additional loop in a new `execute_instructions`.  If the original function had any unit tests, this would have been the natural thing to do.